### DISCUSSION TITLE
Support for GPUs with containerd

### DISCUSSION THREAD

**Discussion by jhosteny** [concourse/concourse#7270] [Ideas]
Hi all,

This was discussed a bit here: https://github.com/concourse/concourse/discussions/6261.

Since then, libcontainerd has been updated to add support for starting GPU-enabled containers. I did a quick spike on this, and was able to build a worker that successfully runs `nvidia-smi`. For now, I hard coded it to add the GPU capability for a privileged task.

The pertinent diff is basically just this:

```patch
diff --git a/worker/runtime/backend.go b/worker/runtime/backend.go
index 9f905dbf2..1c94f7ee7 100644
--- a/worker/runtime/backend.go
+++ b/worker/runtime/backend.go
@@ -237,7 +237,11 @@ func (b *GardenBackend) createContainer(ctx context.Context, gdnSpec garden.Cont
 
        oci.Mounts = append(oci.Mounts, netMounts...)
 
-       return b.client.NewContainer(ctx, gdnSpec.Handle, gdnSpec.Properties, oci)
+       if gdnSpec.Privileged {
+               return b.client.NewGPUContainer(ctx, gdnSpec.Handle, gdnSpec.Properties, oci)
+       } else {
+               return b.client.NewContainer(ctx, gdnSpec.Handle, gdnSpec.Properties, oci)
+       }
 }
 
 func (b *GardenBackend) startTask(ctx context.Context, cont containerd.Container) error {
diff --git a/worker/runtime/libcontainerd/client.go b/worker/runtime/libcontainerd/client.go
index 9cfb1cb16..c555e6e7e 100644
--- a/worker/runtime/libcontainerd/client.go
+++ b/worker/runtime/libcontainerd/client.go
@@ -48,6 +51,17 @@ type Client interface {
                container containerd.Container, err error,
        )
 
+       // NewGPUContainer creates a GPU enabled container in containerd.
+       //
+       NewGPUContainer(
+               ctx context.Context,
+               id string,
+               labels map[string]string,
+               oci *specs.Spec,
+       ) (
+               container containerd.Container, err error,
+       )
+
        // Containers lists containers available in containerd matching a given
        // labelset.
        //
@@ -114,6 +128,20 @@ func (c *client) Stop() (err error) {
        return
 }
 
+func (c *client) NewGPUContainer(
+       ctx context.Context, id string, labels map[string]string, oci *specs.Spec,
+) (
+       containerd.Container, error,
+) {
+       ctx, cancel := createTimeoutContext(ctx, c.requestTimeout)
+       defer cancel()
+
+       return c.containerd.NewContainer(ctx, id,
+               containerd.WithSpec(oci, nvidia.WithGPUs(nvidia.WithDevices(0), nvidia.WithAllCapabilities)),
+               containerd.WithContainerLabels(labels),
+       )
+}
+

```

I'd like to start a discussion on how to add formal support for this. I don't think it would be that hard to do, but I am not sure about the best approach. Would an addition to the [container-limits](https://concourse-ci.org/tasks.html#schema.container_limits) configuration be a logical place for this? Or perhaps a new field on the spec which would map to a reserved property on the Garden container spec?

**Comment by thekyz** [comment]
@jhosteny did you have to do any other configuration on the worker side to make this work? or is it all it takes?

**Reply by jhosteny** [reply]
Hi @thekyz, yes, this is the complete change. However, note that I use the `privileged` flag in the task step to indicate if GPU support is to be enabled. This is done so that check (and other) steps can run normally on the worker. But it will have consequences if you need a non-privileged, non-GPU task to run. In our case, this isn't too much of an issue - we simply will dedicate the machine running this version of the worker to GPU CI tasks via a tag.

A better fix would add specific configuration options to the task step to selectively enable GPU support (and _which_ GPUs should be used, as well). Also, if containerd [eventually supports OCI hooks](https://github.com/containerd/cri/issues/405), that would allow a non-invasive way to support this as well, though I think that it would require the machine to only run the GPU task.

**Reply by thekyz** [reply]
thanks, gonna try that then!

btw your diff is probably missing an import for the "nvidia" spec, or at least I don't see it in the patch

**Reply by thekyz** [reply]
was indeed missing this:
```
diff --git a/worker/runtime/libcontainerd/client.go b/worker/runtime/libcontainerd/client.go
index 9cfb1cb16..0139c12e4 100644
--- a/worker/runtime/libcontainerd/client.go
+++ b/worker/runtime/libcontainerd/client.go
@@ -5,6 +5,7 @@ import (
        "fmt"
        "time"

+       "github.com/containerd/containerd/contrib/nvidia"
        "github.com/containerd/containerd"
        "github.com/opencontainers/runtime-spec/specs-go"
 )
 ```

**Reply by jhosteny** [reply]
@thekyz ah, thanks. Sorry!
